### DISCUSSION TITLE
Can we write custom message hooks in Erlang?

### DISCUSSION THREAD

**Discussion by SSaiPrasad7** [emqx/emqx#4333] [Q&A] [answered]
Hello emqx team,
I am trying to implement hooks using Erlang.I tried the hooks plugins with other languages such as python,lua and java but I want to implement in Erlang.
Is there a way to capture the data with hooks using Erlang(instead of web hook,lua hooks,emqx extension module)?
If the custom Erlang hooks is possible,please let me know how to proceed and end with writing hooks.

**Comment by SSaiPrasad7** [comment]
@zmstone

**Comment by zmstone** [comment, marked as answer]
Hi @saiprasadsamudrala 
We are glad to receive questions like this. It means our effort on supporting community plugin is not useless :)

Yes. We have recently added support for Erlang Plugin development in our new project structure (rebar3 umbrella project in the `master` branch of *this* project).
We have a very brief readme here: https://github.com/emqx/emqx/blob/master/lib-extra
and 'template' plugin for reference, see https://github.com/emqx/emqx/blob/master/lib-extra/plugins

We are also working on Elixir plugin support which will be added to https://github.com/emqx/emqx/blob/master/lib-extra
when it's ready.

We are still working on the documentation updates for the new project structure,
for older versions, you can follow this doc: https://docs.emqx.io/en/broker/latest/advanced/plugins.html
^ which is also worth reading for plugin development in new project structure.

**Reply by SSaiPrasad7** [reply]
Thanks for immediate response @zmstone.
The community version of emqx is pretty good and well documented.Thanks for your support and All the best for upcoming versions of emqx. 
Sure.I will look into documentation.
